src/gateway/Gateway.sol

```diff
@@ -51,47 +73,277 @@ contract Gateway is Auth, IGateway {
         _;
     }
 
+    receive() external payable {
+        emit Received(msg.sender, msg.value);
+    }
+
     // --- Administration ---
-    function file(bytes32 what, address data) public auth {
-        if (what == "poolManager") poolManager = data;
-        else if (what == "aggregator") aggregator = AggregatorLike(data);
-        else if (what == "investmentManager") investmentManager = data;
+    /// @inheritdoc IGateway
+    function file(bytes32 what, address[] calldata routers_) external auth {
+        if (what == "routers") {
+            uint8 quorum_ = uint8(routers_.length);
+            require(quorum_ > 0, "Gateway/empty-router-set");
+            require(quorum_ <= MAX_ROUTER_COUNT, "Gateway/exceeds-max-router-count");
+
+            uint64 sessionId = 0;
+            if (routers.length > 0) {
+                // Increment session id if it is not the initial router setup and the quorum was decreased
+                Router memory prevRouter = activeRouters[routers[0]];
+                sessionId = quorum_ < prevRouter.quorum ? prevRouter.activeSessionId + 1 : prevRouter.activeSessionId;
+            }
+            // Disable old routers
+            for (uint8 i = 0; i < routers.length; i++) {
+                delete activeRouters[routers[i]];
+            }
+
+            // Enable new routers, setting quorum to number of routers
+            for (uint8 j; j < quorum_; j++) {
+                require(activeRouters[routers_[j]].id == 0, "Gateway/no-duplicates-allowed");
+
+                // Ids are assigned sequentially starting at 1
+                activeRouters[routers_[j]] = Router(j + 1, quorum_, sessionId);
+            }
+
+            routers = routers_;
+        } else {
+            revert("Gateway/file-unrecognized-param");
+        }
+
+        emit File(what, routers_);
+    }
+
+    /// @inheritdoc IGateway
+    function file(bytes32 what, address instance) external auth {
+        if (what == "gasService") gasService = GasServiceLike(instance);
+        else if (what == "investmentManager") investmentManager = instance;
+        else if (what == "poolManager") poolManager = instance;
         else revert("Gateway/file-unrecognized-param");
-        emit File(what, data);
+
+        emit File(what, instance);
     }
 
+    /// @inheritdoc IGateway
     function file(bytes32 what, uint8 data1, address data2) public auth {
-        if (what == "message") messages[data1] = data2;
+        if (what == "message") messageHandlers[data1] = data2;
         else revert("Gateway/file-unrecognized-param");
         emit File(what, data1, data2);
     }
 
+    // --- Incoming ---
+    /// @inheritdoc IGateway
```

Comment:
handle has to be `auth` 🙈

Reply:
I think we agreed that it's fine right now. Only routers can call handle. Correct?

Reply:
Yeah I agree this is correct as is.